File: partition-equal-subset-sum/partition-equal-subset-sum.cpp
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int sum=0;
        for(auto it : nums){
            sum+=it;
        }
        
        if(sum%2==1){
            return false;
        }
        std::vector<std::vector<int>> v1(nums.size(),std::vector<int>((sum/2)+1,-1));
        return subsetExists(0,nums,sum/2,v1);
    }
    
    bool subsetExists(int start,std::vector<int>& nums,int target,std::vector<std::vector<int>>& v1){
        if(target==0){
            
            return 1;
        }
        if(start>=nums.size()){
            return 0;
        }
        int include=0;
        
        if(v1[start][target]!=-1){
           return v1[start][target];
        }
        if(target>=nums[start]){
            include=subsetExists(start+1,nums,target-nums[start],v1);    
        }
        int exclude=subsetExists(start+1,nums,target,v1);
        
        v1[start][target]= include or exclude;
        return v1[start][target];
        
    }
};

File: 416-partition-equal-subset-sum/partition-equal-subset-sum.java
class Solution {
    public boolean canPartition(int[] nums) {
        int totalSum = 0;
        for(int num : nums){
            totalSum+=num;
        }
        if(totalSum%2==1){
            return false;
        }
        return subsetSumIterative(nums,totalSum/2);
    }

    boolean subsetSumIterative(int arr[], int target){
        boolean[][] dp = new boolean[arr.length][target+1];
        for(int itr1=0; itr1<arr.length; itr1++){
            dp[itr1][0] = true;
        }
       if (arr[0] <= target) {
        dp[0][arr[0]] = true;
        }
        
        for(int itr1= 1;itr1<arr.length;itr1++){
            for(int itr2=1;itr2<=target;itr2++){
                boolean take = false;
                if(arr[itr1]<=itr2){
                    take = dp[itr1-1][itr2-arr[itr1]];
                }
                boolean notTake = dp[itr1-1][itr2];
                
                dp[itr1][itr2] = take || notTake;
            }
        }
        return dp[arr.length-1][target];
    }
}